lib/api/route/realm.dart

```diff
@@ -0,0 +1,67 @@
+import 'dart:convert';
+
+import 'package:http/http.dart' as http;
+import 'package:json_annotation/json_annotation.dart';
+
+part 'realm.g.dart';
+
+/// https://zulip.com/api/get-server-settings
+///
+/// Despite the name, this is really a home for certain realm-specific
+/// settings, as well as some information about the server as a whole.
+Future<GetServerSettingsResult> getServerSettings({
+  required String realmUrl,
+}) async {
+  // TODO dedupe this part with LiveApiConnection; make this function testable
+  final response = await http.get(
+    Uri.parse("$realmUrl/api/v1/server_settings"));
+  if (response.statusCode != 200) {
+    throw Exception('error on GET server_settings: status ${response.statusCode}');
+  }
+  final data = utf8.decode(response.bodyBytes);
+
+  final json = jsonDecode(data);
+  return GetServerSettingsResult.fromJson(json);
+}
+
+@JsonSerializable(fieldRename: FieldRename.snake)
+class GetServerSettingsResult {
+  final Map<String, bool> authenticationMethods;
+  // final List<ExternalAuthenticationMethod> external_authentication_methods; // TODO handle
+
+  final int zulipFeatureLevel;
+  final String zulipVersion;
+  final String? zulipMergeBase; // TODO(server-5)
+
+  final bool pushNotificationsEnabled;
+  final bool isIncompatible;
+
+  final bool emailAuthEnabled;
+  final bool requireEmailFormatUsernames;
+  final String realmUri;
+  final String realmName;
```

Comment:
`realmName` can be missing, kind of awkwardly. We have this comment in zulip-mobile:

```js
  // When missing, the user requested the root domain of a Zulip server, but
  // there is no realm there. User error.
  //
  // (Also, the server has `ROOT_DOMAIN_LANDING_PAGE = False`, the default.
  // But the mobile client doesn't care about that; we just care that there
  // isn't a realm at the root.)
  //
  // Discussion:
  //   https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1334042
  //
  // TODO(server-future): Expect the error to be encoded in a proper error
  //   response instead of this missing property.
  // TODO(server-future): Then, when all supported servers give that error,
  //   make this property required.
```

and this bit of logic to handle when it's missing:

```js
  if (realm_name == null) {
    // See comment on realm_name in ApiResponseServerSettings.
    //
    // This error copies the proper error that servers *sometimes* give when
    // the root domain is requested and there's no realm there. [1]
    // Specifically, servers give this when
    // `ROOT_DOMAIN_LANDING_PAGE = True`. That circumstance makes no
    // difference to mobile.
    //
    // [1] Observed empirically. For details, see
    //   https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1332900 .
    throw new ApiError(400, { code: 'BAD_REQUEST', msg: 'Subdomain required', result: 'error' });
  }
```

Probably we don't need the runtime code right now, but maybe worth a comment mentioning that there's some complexity to be handled until server-future?

Reply:
Yeah, my thinking was that this result type would describe only what we get when asking at an actual Zulip realm's domain, not a Zulip server's root domain when there's no realm there. I should make that explicit in dartdoc, because the https://zulip.com/api/get-server-settings doc is written with the opposite focus.

When you talk to a server's root domain when no realm is there, you'll get a response that's missing several of these required fields, and so it'll throw an exception. That seems appropriate if one thinks of this as about a realm — other API endpoints don't work on such a domain either. It's basically the same as one expects if the entered URL doesn't belong to a Zulip server at all.

I'll also leave a TODO comment that as part of #35 it would be good to give feedback when the user enters such a domain. I'm not sure we'll necessarily take care of that in the main round of #35 when we close that issue — in particular, I'm not aware of any existing server one can make 
that mistake with other than zulipchat.com, and that one has `ROOT_DOMAIN_LANDING_PAGE = True` so it fails differently — but if not then we can file a more specific issue for it then.